timesheets/Code.gs

```diff
@@ -0,0 +1,148 @@
+      var email = range.getValue();
+      var dropdownValue = sheet.getRange(i,APPROVAL_COLUMN);    
+      
+      // approval email
+      if (dropdownValue.getValue() == 'APPROVED') {
+        var message = 'Your timesheet has been approved';
+        var subject = 'TimeSheet Approval';
+      } else if (dropdownValue.getValue() == 'NOT APPROVED') {
+        var message = 'NOT APPROVED';
+        var subject = 'TimeSheet not Approved';
+      } else if (dropdownValue.getValue() == 'IN PROGRESS') {
+        continue;
+      } 
+      
+      MailApp.sendEmail(email, subject, message);
+      var notifiedUpdate = sheet.getRange(i, lastCol).setValue('NOTIFIED');
+    }
+}
+
+
+/**
+* Runs the "Add Approval Column" script, adding an approval column at the end of the sheet, 
+* made entirely of checkboxes.
+* Adds "NOTIFIED STATUS" column indicating whether or not an employee has yet been e mailed.
+*/
+function addApprovalColumn() { 
+  // adds new column
+  var sheet = SpreadsheetApp.getActiveSheet();
+  var lastCol = sheet.getLastColumn();
+  var lastRow = sheet.getLastRow();
+  sheet.insertColumnAfter(lastCol);
+  
+  // sets new column title
+  var cell = sheet.getRange(1, lastCol+1);
+  cell.setBackground('ffe3ee').setValue('APPROVAL');
+  
+  // make sure approval column is all drop-down menu
+  for (i = 2; i <= lastRow; i++) {
+    var dropdownBox = sheet.getRange(i,lastCol+1);
+    var dropdownValues = ['APPROVED', 'NOT APPROVED', 'IN PROGRESS'];
+    var rule = SpreadsheetApp.newDataValidation().requireValueInList(dropdownValues).build();
+    dropdownBox.setDataValidation(rule);
+    dropdownBox.setValue('IN PROGRESS');
+  }
+  
+  // add the notified column
+  sheet.insertColumnAfter(APPROVAL_COLUMN); //global
+  var cell = sheet.getRange(1, APPROVAL_COLUMN + 1);
+  cell.setBackground('ffe3ee').setValue('NOTIFIED STATUS'); 
+  
+  // sets notified status to NOT NOTIFIED
+  for (var i = 2; i <= lastRow; i++) {
+    var notifiedCell = sheet.getRange(i, APPROVAL_COLUMN + 1);
+    var notifiedValues = ['NOTIFIED', 'NOT NOTIFIED'];
+    var rule = SpreadsheetApp.newDataValidation().requireValueInList(notifiedValues).build();
+    notifiedCell.setDataValidation(rule);
+    notifiedCell.setValue('NOT NOTIFIED');
+  }
+}
+
+/**
+* Runs the "Calculate Pay" script, creating a new column with each employee's 
+* pay based on their reported hours and hourly wage.
+*/
+function calculatePay() {
+  // adds new column
+  var sheet = SpreadsheetApp.getActiveSheet();
+  var lastCol = sheet.getLastColumn();
+  var lastRow = sheet.getLastRow();
+  sheet.insertColumnAfter(lastCol);
+  
+  // sets new column title
+  var cell = sheet.getRange(1, lastCol+1);
+  cell.setValue('WEEKLY PAY');
+  
+  // goes through every employee
+  for (var i = 2; i <= lastRow; i++) {
+    var totalHours = 0;
+    var numHrs = 0;
+    
+    // move across the hour columns
+    for (var j = HRS_START; j <= HRS_END; j++) {
+      var numHrsCell = sheet.getRange(i, j);
```

Comment:
It's not very efficient to use `getRange()` and `getValue()` for each cell in a loop. Better instead to use one `getRange()` and `getValues()` over the entire block of content, and then iterate over the 2D array of values. Ditto for setting values.

This tweet isn't 100% what you need, but gets across the idea:

https://twitter.com/erickoleda/status/1134157734813798400

You should apply this technique to the other functions above as well. Happy to discuss offline as well, as this is more of an advanced tactic.

Reply:
Awesome - this was doable for the cases in which I needed to **get** values. However, I have not been able to figure out how to store an array of ranges (can't find much re: this on the Google Apps Script documentation either), so I have been unable to figure this out for the functions in which I need  **set** values, without calling setValue() every iteration of the loop.